File: frontend/src/api/visualBible.ts
import { apiHeaders, currentApiLocale, parseApiErrorResponse } from './errors'

export type ApprovalStatus = 'draft' | 'approved' | 'archived'
export type VisualEntityType = 'character' | 'outfit' | 'scene' | 'style' | 'prop' | 'control'
export type VisualAssetRole =
  | 'identity_face'
  | 'identity_half_body'
  | 'identity_full_body'
  | 'outfit_front'
  | 'outfit_back'
  | 'outfit_detail'
  | 'scene_master'
  | 'style_reference'
  | 'prop_reference'
  | 'pose'
  | 'depth'
  | 'canny'
  | 'lineart'
  | 'segmentation'
  | 'mask'

export type OutfitVariant = {
  id: number
  project_id: number
  outline_character_id: number
  key: string
  version: number
  name: string
  garment_components: unknown[]
  layer_order: unknown[]
  colors: unknown[]
  materials: unknown[]
  patterns: unknown[]
  accessories: unknown[]
  trigger_tokens: unknown[]
  negative_constraints: string
  status: ApprovalStatus
  approved_at: string | null
  created_at: string
  updated_at: string
}

export type StyleProfile = {
  id: number
  project_id: number
  key: string
  version: number
  name: string
  positive_tag: string
  negative_tag: string
  positive_natural_language: string
  negative_natural_language: string
  color_palette: unknown[]
  lighting: string
  status: ApprovalStatus
  approved_at: string | null
  created_at: string
  updated_at: string
}

export type SceneVisualVersion = {
  id: number
  project_id: number
  script_scene_id: number
  version: number
  landmarks: unknown[]
  spatial_relations: Record<string, unknown>
  camera_presets: unknown[]
  object_states: Record<string, unknown>
  color_palette: unknown[]
  lighting_state: Record<string, unknown>
  status: ApprovalStatus
  approved_at: string | null
  created_at: string
  updated_at: string
}

export type VisualAsset = {
  id: number
  project_id: number
  entity_type: VisualEntityType
  entity_id: number | null
  entity_key: string | null
  role: VisualAssetRole
  storage_kind: 'local_file' | 'renderer_locator'
  local_path: string | null
  renderer_locator: string | null
  mime_type: string | null
  sha256: string | null
  width: number | null
  height: number | null
  version: number
  status: ApprovalStatus
  source: string
  source_image_id: number | null
  crop_metadata: Record<string, unknown>
  mask_asset_id: number | null
  approved_at: string | null
  created_at: string
  updated_at: string
}

const requestJson = async <T>(url: string, options: RequestInit = {}): Promise<T> => {
  const response = await fetch(url, { ...options, headers: apiHeaders(options.headers) })
  if (!response.ok) throw await parseApiErrorResponse(response)
  return (await response.json()) as T
}

export const listOutfits = (projectId: number): Promise<OutfitVariant[]> =>
  requestJson(`/api/visual-bible/projects/${projectId}/outfits`)

export const createOutfit = (
  projectId: number,
  payload: Omit<OutfitVariant, 'id' | 'project_id' | 'version' | 'status' | 'approved_at' | 'created_at' | 'updated_at'>,
): Promise<OutfitVariant> =>
  requestJson(`/api/visual-bible/projects/${projectId}/outfits`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export const listStyles = (projectId: number): Promise<StyleProfile[]> =>
  requestJson(`/api/visual-bible/projects/${projectId}/styles`)

export const createStyle = (
  projectId: number,
  payload: Omit<StyleProfile, 'id' | 'project_id' | 'version' | 'status' | 'approved_at' | 'created_at' | 'updated_at'>,
): Promise<StyleProfile> =>
  requestJson(`/api/visual-bible/projects/${projectId}/styles`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export const listSceneVersions = (projectId: number): Promise<SceneVisualVersion[]> =>
  requestJson(`/api/visual-bible/projects/${projectId}/scene-versions`)

export const createSceneVersion = (
  projectId: number,
  payload: Omit<SceneVisualVersion, 'id' | 'project_id' | 'version' | 'status' | 'approved_at' | 'created_at' | 'updated_at'>,
): Promise<SceneVisualVersion> =>
  requestJson(`/api/visual-bible/projects/${projectId}/scene-versions`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export const setConfigurationStatus = (
  kind: 'outfit' | 'style' | 'scene',
  id: number,
  status: ApprovalStatus,
): Promise<{ id: number; status: ApprovalStatus }> =>
  requestJson(`/api/visual-bible/configurations/${kind}/${id}/status`, {
    method: 'POST',
    body: JSON.stringify({ status }),
  })

export const assignOutfitVariant = (
  scriptCharacterId: number,
  outfitVariantId: number | null,
): Promise<{ id: number; outfit_variant_id: number | null }> =>
  requestJson(`/api/visual-bible/script-characters/${scriptCharacterId}/outfit`, {
    method: 'PUT',
    body: JSON.stringify({ outfit_variant_id: outfitVariantId }),
  })

export const selectSceneVisualVersion = (
  scriptSceneId: number,
  sceneVisualVersionId: number | null,
): Promise<{ id: number; selected_visual_version_id: number | null }> =>
  requestJson(`/api/visual-bible/script-scenes/${scriptSceneId}/visual-version`, {
    method: 'PUT',
    body: JSON.stringify({ scene_visual_version_id: sceneVisualVersionId }),
  })

export const listVisualAssets = (projectId: number): Promise<VisualAsset[]> =>
  requestJson(`/api/visual-bible/projects/${projectId}/assets`)

export const uploadVisualAsset = async (projectId: number, form: FormData): Promise<VisualAsset> => {
  const response = await fetch(`/api/visual-bible/projects/${projectId}/assets/upload`, {
    method: 'POST',
    body: form,
    headers: {
      'X-Locale': currentApiLocale(),
      'Accept-Language': currentApiLocale(),
    },
  })
  if (!response.ok) throw await parseApiErrorResponse(response)
  return (await response.json()) as VisualAsset
}

export const registerVisualAsset = (
  projectId: number,
  payload: {
    entity_type: VisualEntityType
    entity_id: number | null
    entity_key?: string | null
    role: VisualAssetRole
    renderer_locator: string
    sha256?: string | null
    approve: boolean
  },
): Promise<VisualAsset> =>
  requestJson(`/api/visual-bible/projects/${projectId}/assets/register`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export const setVisualAssetStatus = (
  id: number,
  status: ApprovalStatus,
): Promise<VisualAsset> =>
  requestJson(`/api/visual-bible/assets/${id}/status`, {
    method: 'POST',
    body: JSON.stringify({ status }),
  })

export const promoteGeneratedImage = (
  imageId: number,
  payload: {
    entity_type: VisualEntityType
    entity_id: number | null
    entity_key?: string | null
    role: VisualAssetRole
    approve: boolean
  },
): Promise<VisualAsset> =>
  requestJson(`/api/visual-bible/images/${imageId}/promote`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

File: backend/api/schemas/visual_bible.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, model_validator

from backend.models.enums import (
    ApprovalStatus,
    VisualAssetRole,
    VisualEntityType,
)


class OutfitVariantRequest(BaseModel):
    outline_character_id: int = Field(gt=0)
    key: str
    name: str
    garment_components: list[Any] = Field(default_factory=list)
    layer_order: list[Any] = Field(default_factory=list)
    colors: list[Any] = Field(default_factory=list)
    materials: list[Any] = Field(default_factory=list)
    patterns: list[Any] = Field(default_factory=list)
    accessories: list[Any] = Field(default_factory=list)
    trigger_tokens: list[Any] = Field(default_factory=list)
    negative_constraints: str = ""


class OutfitVariantResponse(OutfitVariantRequest):
    id: int
    project_id: int
    version: int
    status: ApprovalStatus
    approved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class StyleProfileRequest(BaseModel):
    key: str
    name: str
    positive_tag: str = ""
    negative_tag: str = ""
    positive_natural_language: str = ""
    negative_natural_language: str = ""
    color_palette: list[Any] = Field(default_factory=list)
    lighting: str = ""


class StyleProfileResponse(StyleProfileRequest):
    id: int
    project_id: int
    version: int
    status: ApprovalStatus
    approved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class SceneVisualVersionRequest(BaseModel):
    script_scene_id: int = Field(gt=0)
    landmarks: list[Any] = Field(default_factory=list)
    spatial_relations: dict[str, Any] = Field(default_factory=dict)
    camera_presets: list[Any] = Field(default_factory=list)
    object_states: dict[str, Any] = Field(default_factory=dict)
    color_palette: list[Any] = Field(default_factory=list)
    lighting_state: dict[str, Any] = Field(default_factory=dict)


class SceneVisualVersionResponse(SceneVisualVersionRequest):
    id: int
    project_id: int
    version: int
    status: ApprovalStatus
    approved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ApprovalRequest(BaseModel):
    status: ApprovalStatus


class VisualAssetLocatorRequest(BaseModel):
    entity_type: VisualEntityType
    entity_id: int | None = Field(default=None, gt=0)
    entity_key: str | None = None
    role: VisualAssetRole
    renderer_locator: str
    sha256: str | None = None
    approve: bool = False

    @model_validator(mode="after")
    def validate_owner(self):
        if self.entity_id is None and not (self.entity_key or "").strip():
            raise ValueError("entity_id or entity_key is required")
        return self


class PromoteImageRequest(BaseModel):
    entity_type: VisualEntityType
    entity_id: int | None = Field(default=None, gt=0)
    entity_key: str | None = None
    role: VisualAssetRole
    approve: bool = False


class VisualAssetResponse(BaseModel):
    id: int
    project_id: int
    entity_type: VisualEntityType
    entity_id: int | None
    entity_key: str | None
    role: VisualAssetRole
    storage_kind: str
    local_path: str | None
    renderer_locator: str | None
    mime_type: str | None
    sha256: str | None
    width: int | None
    height: int | None
    version: int
    status: ApprovalStatus
    source: str
    source_image_id: int | None
    crop_metadata: dict[str, Any]
    mask_asset_id: int | None
    approved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class AssignOutfitRequest(BaseModel):
    outfit_variant_id: int | None = Field(default=None, gt=0)


class SelectSceneVersionRequest(BaseModel):
    scene_visual_version_id: int | None = Field(default=None, gt=0)
